File: frontend/src/lib/api/billing.ts
/**
 * Billing API client for Stripe integration.
 */

import apiClient from './client';

export interface BillingStatus {
  plan_type: string;
  is_paid: boolean;
  paid_at: string | null;
  urls_limit: number | null;
  urls_viewed: number;
  urls_remaining: number | null;
  spots_remaining: number;
  can_upgrade: boolean;
}

export interface CheckoutSession {
  checkout_url: string;
  session_id: string;
}

export interface SpotsRemaining {
  spots_remaining: number;
  max_spots: number;
  spots_claimed: number;
}

/**
 * Get current user's billing status.
 */
export async function getBillingStatus(): Promise<BillingStatus> {
  const response = await apiClient.get<BillingStatus>('/api/v1/billing/status');
  return response.data;
}

/**
 * Create a Stripe checkout session for upgrade.
 */
export async function createCheckoutSession(
  successUrl: string,
  cancelUrl: string
): Promise<CheckoutSession> {
  const response = await apiClient.post<CheckoutSession>('/api/v1/billing/checkout', {
    success_url: successUrl,
    cancel_url: cancelUrl,
  });
  return response.data;
}

/**
 * Get remaining paid spots (public endpoint).
 */
export async function getSpotsRemaining(): Promise<SpotsRemaining> {
  const response = await apiClient.get<SpotsRemaining>('/api/v1/billing/spots-remaining');
  return response.data;
}

File: backend/app/schemas/billing.py
"""
Pydantic schemas for billing endpoints.
"""

from typing import Optional
from pydantic import BaseModel
from datetime import datetime


class CheckoutSessionRequest(BaseModel):
    """Request to create a checkout session."""
    success_url: str
    cancel_url: str


class CheckoutSessionResponse(BaseModel):
    """Response with checkout session URL."""
    checkout_url: str
    session_id: str


class BillingStatusResponse(BaseModel):
    """User's billing/tier status."""
    plan_type: str
    is_paid: bool
    paid_at: Optional[datetime] = None
    urls_limit: Optional[int] = None
    urls_viewed: int = 0
    urls_remaining: Optional[int] = None
    spots_remaining: int
    can_upgrade: bool


class WebhookResponse(BaseModel):
    """Response from webhook processing."""
    status: str
    message: str
